### PR TITLE
=str #16924: Fix FanoutProcessor to not overwrite last termination cause

### PR DESCRIPTION
**Pull request by drewhk**
Also fix FlowSpec to expect onError instead of onComplete for late subscribers

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala [hunk 1]
```diff
@@ -489,11 +489,10 @@ class FlowSpec extends AkkaSpec(ConfigFactory.parseString("akka.actor.debug.rece
...
 
-        // FIXME when adding a sleep before the following link this will fail with IllegalStateExc shut-down
-        // what is the expected shutdown behavior? Is the title of this test wrong?
-        //        val downstream3 = StreamTestKit.SubscriberProbe[Any]()
-        //        publisher.subscribe(downstream3)
-        //        downstream3.expectComplete()
+        val downstream3 = StreamTestKit.SubscriberProbe[Any]()
+        publisher.subscribe(downstream3)
+        downstream3.expectSubscription()
+        downstream3.expectError().getMessage should ===("Cannot subscribe to shut-down Publisher")
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: This message is defined in ActorPublisher.NormalShutdownException. Please extract it to a val and compare with that here. It is also written in StreamOfStreamProcessors.scala and SubscriberManagement.scala. Looks like they could use the ActorPublisher.NormalShutdownException instead.

**Comment by patriknw** [comment]
LGTM, apart from the exception message cleanup

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Ok, thanks!

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala [hunk 2]
```diff
@@ -457,7 +457,7 @@ class FlowSpec extends AkkaSpec(ConfigFactory.parseString("akka.actor.debug.rece
       }
     }
 
-    "call future subscribers' onComplete instead of onSubscribed after initial upstream was completed" in {
+    "call future subscribers' onError after of onSubscribe if initial upstream was completed" in {
```
Comment (2m) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: Delete of.

File: akka-stream/src/main/scala/akka/stream/impl/ActorPublisher.scala [hunk 3]
```diff
@@ -16,8 +16,9 @@ import org.reactivestreams.Subscription
  * INTERNAL API
  */
 private[akka] object ActorPublisher {
-  class NormalShutdownException extends IllegalStateException("Cannot subscribe to shut-down Publisher") with NoStackTrace
-  val NormalShutdownReason: Option[Throwable] = Some(new NormalShutdownException)
+  val NormalShutdownReasonMessage = "Cannot subscribe to shut-down Publisher"
+  class NormalShutdownException extends IllegalStateException(NormalShutdownReasonMessage) with NoStackTrace
+  val NormalShutdownReason: Throwable = new NormalShutdownException
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/ActorPublisher.scala]: The previous value was optimized for not allocating a `Some` during each shutdown, perhaps we can retain this as `val SomeNormalShutdownReason`?

**Comment by rkuhn** [comment]
Other than the deoptimization it LGTM.

**Comment by 2m** [comment]
LGTM apart of a minor type in the test case name.

File: akka-stream/src/main/scala/akka/stream/impl/ActorPublisher.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorPublisher.scala]: Does it matter that much? I expect this not being a hot path. Anyway, the fix is simple.

**Comment by ktoso** [comment]
resolves #16924

**Pull request merged**